geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionEntry.java

```diff
@@ -1338,7 +1338,7 @@ public Object prepareValueForCache(RegionEntryContext r,
           } else {
             data = EntryEventImpl.serialize(val);
           }
-          if (event != null) {
+          if (event != null && event.getCachedSerializedNewValue() == null) {
```

Comment:
I don't like this. This check was already done. So just change this code like so:
if (isSerialized) {
  if (event != null && event.getCachedSerializedNewValue() == null) {
    data = event.getCachedSerializedNewValue();
  } else {
    // compute serialized value
    if (CachedSerializable) ...
    else if (PdxInstance) ...
    else serialize it
    if (event != null) {
      event.setCachedSerializedNewValue(data);
    }
  }
}

Reply:
With your suggested code changes,
if (isSerialized) {
if (event != null && event.getCachedSerializedNewValue() == null) {
data = event.getCachedSerializedNewValue(); // This will keep data as null and NullPointerException using it in allocateAndInitialize.
} else {
// compute serialized value
if (CachedSerializable) ...
else if (PdxInstance) ...
else serialize it
if (event != null) {
event.setCachedSerializedNewValue(data);
}
}
}

And with the code updated in this PR
        if (isSerialized) {
          if (event != null && event.getCachedSerializedNewValue() != null) {
            data = event.getCachedSerializedNewValue();
          } else if (val instanceof CachedDeserializable) {
            data = ((CachedDeserializable) val).getSerializedValue();
          } else if (val instanceof PdxInstance) {
            try {
              data = ((ConvertableToBytes) val).toBytes();
            } catch (IOException e) {
              throw new PdxSerializationException("Could not convert " + val + " to bytes", e);
            }
          } else {
            data = EntryEventImpl.serialize(val);
          }
          if (event != null  && event.getCachedSerializedNewValue() == null) { // This will prevent an unneeded op as you suggested earlier.
            event.setCachedSerializedNewValue(data);
          }
        } else {
          data = (byte[]) val;
        }


I see this check is OK before setting the new value which prevents a no op as well as caches the value in the event.

if (event != null && event.getCachedSerializedNewValue() == null) { 
  event.setCachedSerializedNewValue(data);
}